src/main/java/me/moodcat/backend/rooms/RoomInstance.java

```diff
@@ -206,8 +217,11 @@ private void processVotes(final Song previousSong) {
     private void processNextSong(final Room room, final List<Song> history) {
         final List<Song> playQueue = room.getPlayQueue();
         if (playQueue.isEmpty()) {
-            playQueue.addAll(history);
-            history.clear();
+            VAVector vector = this.roomDAOProvider.get().findById(id).getVaVector();
+            playQueue.addAll(this.songDAOProvider.get().findForDistance(vector, NUMBER_OF_SELECTED_SONGS));
+            
+            int historyLength = history.size();
+            room.setPlayHistory(history.subList(Math.max(0, historyLength - NUMBER_OF_SELECTED_SONGS), historyLength));
```

Comment:
The last index is exclusive, so when we take the length of the list, the last item included is `length - 1`. If we would decrement the lastindex with another, the last item included is `length - 1 - 1`.